### ISSUE TITLE
barcode and umi opposite sides and whitelist optimizations

### ISSUE THREAD

**Issue by ljwharbers** [DavidsonGroup/flexiplex#44]
Hi,

First off, great tool! This is exactly what I was looking for.

I have two questions, the first is regarding how to run flexiplex when your barcode and UMI are on different sides. 
My read structure is as follows: 
> Adaptor_seq1 -- BC (32bp) -- spacer_seq -- polyA -- cDNA -- UMI -- Adaptor_seq2

Do I understand it correctly that the best way to do this is as follows:
`gunzip -c input.fastq.gz | flexiplex -x CTACACGACGCTCTTCCGATCT -b "???????????????????????????????" -u "" -f 4 | flexiplex -x TTTTTTTT -b "" -u "????????" -x CAGACGTGTGCTCTTCCGATCT -f 4`

My second question arises from the fact that my whitelist is millions of barcodes (determined with short read illumina sequencing). The combination of the long barcodes and the millions of whitelisted barcodes will make the barcode matching impossible. I tried to run it on a smaller subset of my data initially in discovery mode and obtain a smaller whitelist using the flexiplex_filter script, but this will still result in >200K barcodes (likely still over a million with the full dataset). Do you have any suggestions or possible workarounds. Potentially only searching for perfect matches with the whitelist would be a solution, but I don't think this is currently implemented(?).

Best,
Luuk

**Comment by olliecheng** [comment]
Hi! Thanks for your kind words.

At the moment, your first flexiplex command is in 'discovery' mode and so won't work when piped directly into another run, as it doesn't produce a .fastq output - instead, it produces a table of barcode counts. However, if you had a whitelist passed in with `-k`, you would be able to pipe it in just fine.

When it comes to your second run of Flexiplex, the structure that the tool is looking for reads from left to right, with the adapter/BC/UMI sequence on the left and the cDNA on the right. This isn't an issue in practice as Flexiplex scans both the forward and reverse complement, but it does mean that if you expect something like:
```
cDNA + "TTTTTTTT" + umi + "CAGACGTGTGCTCTTCCGATCT"
```
in your forward strand, to demultiplex with Flexiplex you would want to search for the reverse complement of that:
```
AGATCGGAAGAGCACACGTCTG + umi_rev_cmp + "AAAAAAAA" + cDNA
```
Furthermore, as per the [docs](https://davidsongroup.github.io/flexiplex/#extracting-umis-from-pcr-cdna-ont-data), if you only want to extract the UMI and know that there should be no barcode, you would also pass in `-b "" -k "?"`. So, I think the correct usage of Flexiplex in this situation would be:
```
flexiplex -x "AGATCGGAAGAGCACACGTCTG" -u "????????" -x "TTTTTTTT" -b "" -k "?" -f 4
```
You may find it helpful to use the `-n` flag to set different prefixes for the output files produced by both stages, otherwise they'll start overwriting each other.

As for your second question, you should be able to configure the barcode error tolerance to be zero with the `-e 0` flag, except if you have a high number of barcodes this would still be pretty slow. Alternatively, if you'd prefer not to set a barcode list, I think `-k "?"` should work... kind of. It will allow anything which matches the format of `-b`, but I don't think the read header contains the actual barcode as a result. This information may be able to be extracted from the `flexiplex_reads_barcodes.txt`, though... Is the ideal solution for you to have an option to extract barcodes without a whitelist (as if the whitelist contained every possible kmer I guess), so you can clean up the data downstream?

Cheers
OIllie

**Comment by olliecheng** [comment]
I should add as an aside: just earlier today, I had a colleague ask me the same question about a BC + cDNA + UMI situation! It might be a good idea to add this to the documentation...

**Comment by nadiadavidson** [comment]
Hi Luuk,

Thanks for sending so much detail about your use case!

As a simple work around for your first issue of not having the barcodes reported. I think you could treat the barcodes as UMIs. So run something like:
gunzip -c input.fastq.gz | flexiplex -x CTACACGACGCTCTTCCG -b "ATCT" -u "???????????????????????????????" -f 3 -e 1 -k "ATCT" -n barcode | flexiplex -x "AGATCGGAAGAGCACACGTCTG" -u "????????" -x "?" -b "" -k "?" -f 4 -n umi > test.fastq
Then you'll get the the barcode in the read, prefaces with the ATCT_ which hopefully is easy to ignore in your down stream analysis. 

However, this won't do any error correction of the barcodes using the whitelist, which is essential your second issue/question. If you would like to error correct to the Ilumina list, my only thought would be to spit your reads up into multiple subfiles and run them in parallel, providing the whitelist to the -k parameter. Flexiplex has an option for multithreads as well, which you could try, but this tends to be slower than file splitting for large threads in our experience. Either way, my guess it that it will be very slow with millions of barcodes which are 32bp long.

This is a really nice scenario for us to think about how we can improve the performance of the tool. So thanks for bringing it to our attention!

All the best,
Nadia.

**Comment by ljwharbers** [comment]
Hi Nadia,

Unfortunately we just missed eachother during your visit to Leuven!

Thanks for the response and workaround suggestion! This does indeed seem to work perfectly and it shouldn't be a problem to extract the barcode in this way.

Regarding the error correction, I have an implementation to do 1 mismatch correction (written in c++), but in this particular dataset it does not rescue many read. Correcting using the levenshtein distance would of course be better but computationally not feasible for this dataset. That's why, for the moment, I'm not too bothered to perform the mismatch correction. 

I'm definitely going to use this approach over my current implementation, a hacked variant of blaze, since it seems to perform a lot better compared to my hacked variant of blaze (both in spee, number of reads it retains, and simplicity of running it).

Thanks again and I'm more than happy to help testing things if you're looking to improve the performance on datasets like mine.

Best,
Luuk

**Comment by nadiadavidson** [comment]
Ah I hadn't realised you are from Leuven! It's a wonderful city and institute!
We are very happy to help, so let us know how everything goes and if you have any more questions or requests.

Cheers,
Nadia.
